test/spmd/tensor/test_dtensor_ops.py

```diff
@@ -0,0 +1,875 @@
+c32 = torch.complex32
+c64 = torch.complex64
+c128 = torch.complex128
+i8 = torch.int8
+i16 = torch.int16
+i32 = torch.int32
+i64 = torch.int64
+b8 = torch.bool
+u8 = torch.uint8
+
+dtype_abbrs = {
+    torch.bfloat16: "bf16",
+    torch.float64: "f64",
+    torch.float32: "f32",
+    torch.float16: "f16",
+    torch.complex32: "c32",
+    torch.complex64: "c64",
+    torch.complex128: "c128",
+    torch.int8: "i8",
+    torch.int16: "i16",
+    torch.int32: "i32",
+    torch.int64: "i64",
+    torch.bool: "b8",
+    torch.uint8: "u8",
+}
+
+# rewrite common size variables to sth can be sharded evenly
+# we can enable uneven shards later, but need to adjust more on
+# sample inputs (i.e. view/reshape need to adjust shape size as well)
+common_ops.L = 24
+common_ops.M = 12
+common_ops.S = 4
+common_ops.XS = 2
+
+# torch.testing._internal.common_method_invocations.
+
+
+def assert_ref_dtensor_equal(test_case, dtensor_rs, rs, msg_callable):
+    mesh = test_case.mesh
+    flat_dtensor_rs, _ = tree_flatten(dtensor_rs)
+    flat_rs, _ = tree_flatten(rs)
+    test_case.assertEqual(len(flat_dtensor_rs), len(flat_rs))
+    for i, dtensor_r, r in zip(range(len(flat_rs)), flat_dtensor_rs, flat_rs):
+
+        def test_assert(cond, msg):
+            if not cond:
+                raise RuntimeError(f"output {i}: {msg_callable(msg)}")
+
+        if not isinstance(r, torch.Tensor):
+            continue
+        test_assert(
+            isinstance(dtensor_r, torch.Tensor),
+            f"but real {i}th result is Tensor",
+        )
+        test_assert(dtensor_r.dtype == r.dtype, f"but real dtype was {r.dtype}")
+        test_assert(dtensor_r.shape == r.shape, f"but real shape was {r.shape}")
+        # NOTE: stride checking is currently disabled
+        # See https://github.com/pytorch/pytorch/issues/78050
+        # same_strides, _ = prims.utils.check_significant_strides(meta_r, r)
+        # test_assert(same_strides, f"but real stride was {r.stride()}")
+        test_assert(
+            dtensor_r.requires_grad == r.requires_grad,
+            f"but real requires_grad was {r.requires_grad}",
+        )
+
+        # redistribute/all_gather the results to compare with normal output
+        full_out = dtensor_r.redistribute(
+            mesh, mesh.ndim * [Replicate()]
+        ).to_local()
+        test_case.assertEqual(full_out, r)
+
+
+# This environment variable controls whether or not we print expected failure
+# lists at the end of a test suite run.  The intended usage looks like this:
+#
+# 1. Run `PYTORCH_COLLECT_EXPECT=1 python test/test_dtensor_ops.py` on a CUDA build
+#    of PyTorch that has LAPACK/MAGMA installed.
+# 2. Given the printed skip/xfail list, add them to the corresponding lists;
+#    torch.* entries go in meta_function and aten.* entries go in meta_dispatch.
+#    If there are preexisting entries, you need to merge in the entries.
+#
+# This is somewhat manual but typically you shouldn't need to do this, unless
```

Comment:
Do you mean this is just a manual test?

Reply:
No i mean this environment variable could be used to generate the failed list and we could copy them 1 by 1, normally we don't set this environment variable, but I leave it here in case in the future we want to generate failed list again